src/rust/inetstack/protocols/mod.rs

```diff
@@ -13,7 +13,34 @@ pub mod udp;
 
 pub use peer::Peer;
 
+use std::slice::ChunksExact;
+
 pub enum Protocol {
     Tcp,
     Udp,
 }
+
+pub fn generic_checksum(buf: &[u8], start: Option<u32>) -> u32 {
+    let mut state: u32 = match start {
+        Some(state) => state,
+        None => 0xFFFF
+    };
+
+    let mut chunks_iter: ChunksExact<u8> = buf.chunks_exact(2);
+    while let Some(chunk) = chunks_iter.next() {
+        state += u16::from_be_bytes([chunk[0], chunk[1]]) as u32;
+    }
+
+    if let Some(&b) = chunks_iter.remainder().get(0) {
+        state += u16::from_be_bytes([b, 0]) as u32;
+    }
+
+    state
+}
+
+pub fn fold16(start: u32) -> u16 {
```

Comment:
Could you add a rust doc here that explains what this function does?

Reply:
Done.